File: linkedlist_cycle.py
class Node:
  def __init__(self, value, next=None):
    self.value = value
    self.next = next


def has_cycle(head):
  hare = head
  tortoise = head
  while hare is not None and tortoise is not None:
    hare = hare.next.next
    tortoise = tortoise.next
    if hare == tortoise:
      return True
  return False

# O(N) time complexity, where N is the length of the given linkedlist, since the hare will meet the tortoise in the same loop.
# O(1) space complexity, since we are not creating any new data structures.

File: Java/linkedlist_cycle.java
class ListNode {
  int value = 0;
  ListNode next;

  ListNode(int value) {
    this.value = value;
  }
}

class LinkedListCycle {

  public static boolean hasCycle(ListNode head) {

    if (head == null) {
      return false;
    }

    ListNode fast = head;
    ListNode slow = head;

    while (fast != null) {
      fast = fast.next.next;
      slow = slow.next;

      if (fast == slow) {
        return true;
      }
    }
    return false;
  }
